From 455579239e7c8f1793c70f51e7f03aaa8ef24676 Mon Sep 17 00:00:00 2001
From: Tamizh Chelvam Raja <tamizhr@qti.qualcomm.com>
Date: Mon, 26 May 2025 21:55:10 +0530
Subject: [PATCH] UPSTREAM:wifi:ath12k: Fix possible NULL access

In ath12k tx completion the vif is taken from skb_cb->vif
and that can become NULL in interface removal case.
To avoid this NULL crash do not give any extra room
between skb_cb->vif NULL check and its usage.

Signed-off-by: Tamizh Chelvam Raja <tamizhr@qti.qualcomm.com>
---
 drivers/net/wireless/ath/ath12k/dp_tx.c | 12 ++++++------
 1 file changed, 6 insertions(+), 6 deletions(-)

--- a/drivers/net/wireless/ath/ath12k/dp_tx.c
+++ b/drivers/net/wireless/ath/ath12k/dp_tx.c
@@ -908,12 +908,12 @@ ath12k_dp_tx_htt_tx_complete_buf(struct
 
 	skb_cb = ATH12K_SKB_CB(msdu);
 
-	if (unlikely(!skb_cb->vif)) {
+	vif = skb_cb->vif;
+	if (unlikely(!vif)) {
 		ieee80211_free_txskb(ar->ah->hw, msdu);
 		return;
 	}
 
-	vif = skb_cb->vif;
 	ahvif = (void *)vif->drv_priv;
 	link_id = skb_cb->link_id;
 	if (ahvif->links_map & BIT(link_id)) {
@@ -1181,15 +1181,15 @@ static void ath12k_dp_tx_complete_msdu(s
                 return;
         }
 
-        if (unlikely(!skb_cb->vif)) {
-                ieee80211_free_txskb(ar->ah->hw, msdu);
-                return;
-        }
-
 	ts.buf_rel_source = buf_rel_source;
 	ath12k_dp_tx_status_parse(ab, tx_status, &ts);
 
 	vif = skb_cb->vif;
+        if (unlikely(!vif)) {
+                ieee80211_free_txskb(ar->ah->hw, msdu);
+                return;
+        }
+
 	ahvif = (void *)vif->drv_priv;
 	link_id = skb_cb->link_id;
 
